From a86a429fe59f8f8386771f7aefa0cb4cc68e4897 Mon Sep 17 00:00:00 2001
From: _M_z <hardirq@outlook.com>
Date: Thu, 6 Feb 2025 22:43:08 +0800
Subject: [PATCH] feat: /run/pppd/resolv.conf

To avoid read-only system in NixOS.
---
 pppd/pathnames.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/pppd/pathnames.h b/pppd/pathnames.h
index ddc5e76d..a8aac6ef 100644
--- a/pppd/pathnames.h
+++ b/pppd/pathnames.h
@@ -101,7 +101,7 @@
 #define PPP_PATH_AUTHDOWN       PPP_PATH_CONFDIR "/auth-down"
 #define PPP_PATH_TTYOPT         PPP_PATH_CONFDIR "/options."
 #define PPP_PATH_PEERFILES      PPP_PATH_CONFDIR "/peers/"
-#define PPP_PATH_RESOLV         PPP_PATH_CONFDIR "/resolv.conf"
+#define PPP_PATH_RESOLV         "/run/pppd/resolv.conf"
 
 #define PPP_PATH_NET_INIT	PPP_PATH_CONFDIR "/net-init"
 #define PPP_PATH_NET_PREUP	PPP_PATH_CONFDIR "/net-pre-up"
